CRM/Contribute/BAO/Contribution.php

```diff
@@ -1725,7 +1725,19 @@ public static function transitionComponents($params, $processContributionObject
     if ($contributionStatusId == array_search('Cancelled', $contributionStatuses)) {
       if (is_array($memberships)) {
         foreach ($memberships as $membership) {
-          if ($membership) {
+          $update = TRUE;
+          //Update Membership status if there is no other completed contribution associated with the membership.
+          $relatedContributions = CRM_Member_BAO_Membership::getMembershipContributionId($membership->id, TRUE);
+          foreach ($relatedContributions as $contriId) {
+            if ($contriId == $contributionId) {
+              continue;
+            }
+            $status = CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Contribution', $contriId, 'contribution_status_id');
```

Comment:
Can we used a cached function here? - ie

if (CRM_Core_PseudoConstant::getName('CRM_Contribute_BAO_Contribution', 'contribution_status_id', $contributionStatusId) === 'Completed')